Restructure the userdoc directory

jessica-mitchell opened the pull request:
This PR reorganizes the documentation to improve findability of different topics, both in the source on GitHub and on Read the docs.

In the userdoc/ directory the guides/ have been removed in favour of more descriptive folders
New folders:

- neuron_docs/  -- docs about neurons
- synapse_docs/ -- docs about synapses and connections
- device_docs/ -- docs about devices
- coupling_nest_docs/  -- docs about other tools that connect with nest
- nest_behavior_docs/ -- docs about how nest works
- hpc_docs/ -- docs about hpc setup

A new table of contents was also created for highlighting topics on Read the Docs:
You can see the structure output here: https://nest-test.readthedocs.io/en/new-doc-folder-structure/contents.html

In addition, I moved the developer docs back into userdoc/ because
1. I want to build the dev docs with Sphinx and that's done easiest if it's in userdoc/
2. I want to include all docs that relate to maintaining, contributing or developing the documentation or code in anyway in devdoc/.  (Thus all contribution guidelines, styleguides, workflows, go in here) 
3. Note that the actual doxygen output is not rendered on Rtd yet

See the output table of contents for the developer area here: https://nest-test.readthedocs.io/en/new-doc-folder-structure/devdoc/index.html

Please comment if you have suggestions for alternatives!

clinssen commented:
- In general, I'm not super enthusiastic about making subdirectories ``userdoc`` and ``devdoc``, because these "deep" hierarchies never work in practice. Now this PR has ``doc/userdoc/devdoc``, which is particularly confusing. I would suggest to put everything just right under ``doc`` in the root of the repository.

- > A new table of contents was also created for highlighting topics on Read the Docs

  This is just to illustrate the new menu structure, right? The items are one-to-one the same as on the left-hand column menu. I'd recommend against including the page itself (a site should not need a "sitemap").

- Do we need the ``_docs`` suffixes on all subdirectories?

jessica-mitchell commented:
> * In general, I'm not super enthusiastic about making subdirectories `userdoc` and `devdoc`, because these "deep" hierarchies never work in practice. Now this PR has `doc/userdoc/devdoc`, which is particularly confusing. I would suggest to put everything just right under `doc` in the root of the repository.

Yes, I agree to some extent - I would not want devdoc under userdoc; this was just a way to get the docs to build with the current hierarchy.
I do think we can separate developer docs from user docs - in the sense that I mentioned above: any documentation that is used to  modify and develop code or documentation, all workflows (for contributing, git etc) and templates should be under 'developer'.
We could also remove devdoc / userdoc completely and create subdirectories for each topic. But I like the idea of grouping all docs that a developer may need together, as is here https://nest-test.readthedocs.io/en/new-doc-folder-structure/devdoc/index.html. 
 But this could be done via linking to different folders in the repo.
> 
>     * > A new table of contents was also created for highlighting topics on Read the Docs
>       
>       
>       This is just to illustrate the new menu structure, right? The items are one-to-one the same as on the left-hand column menu. I'd recommend against including the page itself (a site should not need a "sitemap").

Yes, I showed the toc page to highlight the structure. But this page is available on Read the  docs (currently if you click the nest logo I think you get contents.html, automatically). It is what the sidebar will show on every page, including the new theme.

> 
>     * Do we need the `_docs` suffixes on all subdirectories?

Probably not - I am up for removing it, and any alternative naming suggestions.

jessica-mitchell commented:
@pnbabu @clinssen After some discussions with @jougs and @steffengraber, we decided to keep the hierarchy (doc/ userdoc/) to ensure nest can build on Read the docs and locally. I changed the name userdoc to htmldoc, and updated the file organization as we discussed in hackathon. see here: https://nest-test.readthedocs.io/en/new-doc-folder-structure/contents.html

pnbabu commented on doc/htmldoc/developer_space/index.rst:
```diff
@@ -0,0 +1,53 @@
+   :caption: SLI docs
+   :glob:
+
+   sli_docs/*
```
This shows all the links under `sli_doc` and also a link to that page. Should the `*` be removed here?

jessica-mitchell replied:
fixed

pnbabu commented on doc/htmldoc/models/models-main.rst:
```diff
@@ -0,0 +1,13 @@
+
+Need to create your own?
+
+* Check out :doc:`NESTML <nestml:index>`, a domain-specific language supporting neuron and synapse models without
+  any C++.
```
```suggestion
* Check out :doc:`NESTML <nestml:index>`, a domain-specific language supporting neuron and synapse models with a code generation backend that generates model code in C++ for NEST.
```

pnbabu reviewed (commented):
I observed these inconsistencies while browsing through the doc:
- Images in `parameterization.rst` are not displayed
- The hrefs inside `handling_connections.rst`, `spatially_structured_networks,rst` don’t work

jessica-mitchell commented on doc/htmldoc/contents.rst:
```diff
@@ -0,0 +1,64 @@
+   Neurons <neurons/index>
+   Synapses and connections <synapses/index>
+   Devices  <devices/index>
+   Spatially-structured networks <networks/index>
```
Good point, fixed

jessica-mitchell commented on doc/htmldoc/contents.rst:
```diff
@@ -0,0 +1,64 @@
+   Simulation behavior <nest_behavior/running_simulations>
+   Random numbers <nest_behavior/random_numbers>
+   Built-in timers <nest_behavior/built-in_timers>
+   Connect to other tools <nest_behavior/index>
```
Yes

jessica-mitchell commented:
@pnbabu thanks for your comments; I believe I fixed them all

clinssen commented:
Should ``extractor_userdocs.py`` and the directory that it creates also be renamed?

pnbabu commented:
A general question: clicking on the NEST logo on the docs page takes to "Table of Contents". Is this behaviour intended? Shouldn't this point to the NEST home page instead?

jessica-mitchell commented:
> Should `extractor_userdocs.py` and the directory that it creates also be renamed?

I don't think so, because this extracts the "Begin Userdoc" section from the .h model files. So imo, this is a good use of the word userdocs.

> A few thoughts about the left-hand side menu:
> 
> "HPC optimization" has only one sub-element, collapse the hierarchy for now and just link directly to "Guide to parallel computing".

Ok set it to parallel computing. 

> "Release notes" can probably go under "reference material", it doesn't need to be such a prominent item in the menu.

done.
> "Simulation behaviour" page title is "How to run simulations"; make consistent. Same for "Randomness in NEST Simulations", "Coupling NEST with other tools", "Contact us".

Titles/links updated 
> "Command-line help" actually links to "Getting help" page.

updated the links and made some slight modifications to content to be more clear
> 
> "News and updates" and a few more links go outside of RtD. Could you mark these with a special icon, and/or make them open in _blank?

I couldn't get an icon inserted in the toc, and I didnt find a simple solution for opening the pages in new tabs. I'd have to look into this more in depth. Maybe something for a future PR if that works for you.

jessica-mitchell commented:
> A general question: clicking on the NEST logo on the docs page takes to "Table of Contents". Is this behaviour intended? Shouldn't this point to the NEST home page instead?

This is default behavior of this theme,  because we have contents.rst (the table of contents) as the `master_doc` in the conf.py.  (note nestml's  `master_doc = index`.)  Also, we can't change the `master_doc` to index, without causing issues with sidebars.

jessica-mitchell commented:
@pnbabu @clinssen I think I have addressed all you comments thus far. Just a kind reminder to finish the review when you can, thanks!

pnbabu reviewed (approved):
@jessica-mitchell The CI is failing for something unrelated to the changes in this PR. Otherwise, this looks good to me.

clinssen commented:
Inspired by https://documentation-for-annarchy.readthedocs.io/, can we have a menu that only has the items:
* **Home**: subsumes current "Cite NEST" and "License". Add a link to "NEST Initiative" on the front page with a logo; external link should not be in the menu structure. Same for "Publications". Also subsume "Related Projects". All of these can just be rendered as different headers on a single landing page. Front page could include the video tutorial.
* **Installation**: current "Install NEST"
* **Manual**: covers NEST internals, current "Understand how NEST works", "Glossary", "From NEST 2.x to 3.x", "Model directory", MUSIC interface
* **Examples**: subsumes current PyNEST part of "Tutorials" and "PyNEST example scripts"
* **API**: current "PyNEST API", hopefully in the future this will also include the C++ API
* **Contributing to NEST**: subsumes current "Developer space"
* **Getting help**: subsumes current "Contact us" and "Get help"

By "subsume", I mean these pages go under this part of the hierarchy.

All of these top-level menu items would be collapsed by default, giving some space rather than the overwhelming number of items there are now.

jessica-mitchell commented:
@clinssen I've simplified the main toc based on our discussion. I kept the model directory as a link, becasue I thought that might be something like the API users may want to explore.  Please approve if you agree :) 
Just something to note for future, I think there is more work to be done to make index pages more visually appealing, but I think the structure is vastly improved.

jessica-mitchell commented:
@clinssen ping!

clinssen commented:
This is shaping up really nicely! Just a few small points.

- Move "From NEST 2.x to 3.x" to the bottom of the list under "Understand how NEST works"
- "NEST models" page can probably be removed; links should be merged elsewhere where it makes sense
- "Troubleshooting" is not in any category, could go under "Get help"
- "Get help" and "Community" seems like they could be merged. I anyway expect nest.help() and nest.helpdesk() to disappear before long, which eliminates most of the content of the "Get help" page.
- "Community" page: first sections are links, later sections are text on the page; this is confusing. 
- Can we add an icon in front of RtD-external links? See ``style_external_links`` option in the template.
- Clicking the top-left logo should link to the landing page, not ToC
- GNU License link on the front page is broken
- ``padding-left: 2rem`` seems excesive on unordered lists, especially the left-side menu column. Please make different styles for ``wy-menu ul`` and ``rst-content ul``.

clinssen commented:
Many thanks! I'm approving already.

Minor things to fix:
- NEST Contributor Agreement link is broken on the Contribute page.
- FAQ seems to be missing in the menu (perhaps I overlooked it). Perhaps under Community/Get Help? Or should FAQ be removed altogether as it's a bit of a "miscellaneous" category?

Perhaps you could double-check the RtD documentation build log; I'm not sure but it might emit warnings there in case of broken links or orphaned documents.

jessica-mitchell commented:
@clinssen thanks for approving I will take note of your last comments and apply them to a different PR
